### PR TITLE
.d.ts files don't have outputs

### LINKED ISSUE #25355: tsc --build mode: "<package> is out of date", when it is up-to-date
**TypeScript Version:**  3.0.0-dev.20180630

**Search Terms:** 
build mode, out of date, incremental

**Steps to Reproduce**

Set up @RyanCavanaugh's [`learn-a` sample repo](https://github.com/RyanCavanaugh/learn-a) as per his [instructions](https://github.com/RyanCavanaugh/learn-a#setting-up-this-repo). Run `tsc -b packages --verbose` to see everything gets built the first time. Now add an empty declaration file at `pkg1/src/bar.d.ts`. Run `tsc -b packages --verbose` several more times in succession.

**Expected behavior:**

Build for `pkg1` should be skipped since its output is up-to-date. `pkg1/src/bar.d.ts` is never going to generate any output.

**Actual behavior:**

On every build, `tsc` reports `Project 'packages/pkg1/tsconfig.json' is out of date because output file 'packages/pkg1/lib/bar.js' does not exist`, and re-builds `pkg1`.

**Related Issues:**

#3469

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #25355

### CODE REVIEW

File: src/compiler/tsbuild.ts [hunk 1]
```diff
@@ -294,8 +294,10 @@ namespace ts {
         return changeExtension(outputPath, (fileExtensionIs(inputFileName, Extension.Tsx) && configFile.options.jsx === JsxEmit.Preserve) ? Extension.Jsx : Extension.Js);
     }
 
+    const extensionsWithNoOutput: ReadonlyArray<string> = [Extension.Dts, Extension.Json];
     function getOutputFileNames(inputFileName: string, configFile: ParsedCommandLine): ReadonlyArray<string> {
-        if (configFile.options.outFile) {
+        // outFile is handled elsewhere; .d.ts files don't generate outputs
+        if (configFile.options.outFile || fileExtensionIsOneOf(inputFileName, extensionsWithNoOutput)) {
```
Comment (mhegazy) [inline src/compiler/tsbuild.ts]: `options.out` as well.

Comment (mhegazy) [reply src/compiler/tsbuild.ts]: also what about `--noEmit`?

File: src/compiler/tsbuild.ts [hunk 2]
```diff
@@ -294,8 +294,10 @@ namespace ts {
         return changeExtension(outputPath, (fileExtensionIs(inputFileName, Extension.Tsx) && configFile.options.jsx === JsxEmit.Preserve) ? Extension.Jsx : Extension.Js);
     }
 
+    const extensionsWithNoOutput: ReadonlyArray<string> = [Extension.Dts, Extension.Json];
     function getOutputFileNames(inputFileName: string, configFile: ParsedCommandLine): ReadonlyArray<string> {
-        if (configFile.options.outFile) {
+        // outFile is handled elsewhere; .d.ts files don't generate outputs
```
Comment (sheetalkamat) [inline src/compiler/tsbuild.ts]: Json files are emitted just like js files. But they do not have sourcemaps.

Comment (weswigham) [reply src/compiler/tsbuild.ts]: Really? So if I `require`/`import` my package.json it'll bet included in the emit? o.O

Comment (RyanCavanaugh) [reply src/compiler/tsbuild.ts]: Count me as surprised as well, but verified and this is the case

**Comment by StevenLangbroek** [comment]
Interesting note: I had missed the `esModuleInterop` param from the announcement, and adding it actually made json imports work for me. Is that expected? (2.9.1)

**Pull request merged**